thetis/solver.py

```diff
@@ -71,29 +71,75 @@ def __init__(self, mesh2d, bathymetry_2d, n_layers,
         """Do export initial state. False if continuing a simulation"""
         self._isfrozen = True  # disallow creating new attributes
 
-    def set_time_step(self):
+    def compute_time_step_2d(self, u_mag=Constant(0.0)):
+        """
+        Computes maximum explicit time step from CFL condition.
+
+        dt = CellSize/U
+
+        Assumes velocity scale U = sqrt(g*H) + u_mag
+        where u_mag is estimated advective velocity
+        """
+        csize = self.fields.h_elem_size_2d
+        bath = self.fields.bathymetry_2d
+        fs = bath.function_space()
+        bath_pos = Function(fs, name='bathymetry')
+        bath_pos.assign(bath)
+        min_depth = 0.05
+        bath_pos.dat.data[bath_pos.dat.data < min_depth] = min_depth
+        test = TestFunction(fs)
+        trial = TrialFunction(fs)
+        solution = Function(fs)
+        g = physical_constants['g_grav']
+        u = (sqrt(g * bath_pos) + u_mag)
+        a = inner(test, trial) * dx
+        l = inner(test, csize / u) * dx
+        solve(a == l, solution)
+        return solution
+
+    def compute_time_step_3d(self, u_mag=Constant(1.0)):
+        """
+        Computes maximum explicit time step from CFL condition.
+
+        dt = CellSize/U
+
+        Assumes velocity scale U = u_mag
+        where u_mag is estimated advective velocity
+        """
+        csize = self.fields.h_elem_size_2d
+        bath = self.fields.bathymetry_2d
+        fs = bath.function_space()
+        test = TestFunction(fs)
+        trial = TrialFunction(fs)
+        solution = Function(fs)
+        a = inner(test, trial) * dx
+        l = inner(test, csize / u_mag) * dx
+        solve(a == l, solution)
+        return solution
+
+    def set_time_step(self, alpha=0.05):
         comm = self.comm
         if self.options.use_mode_split:
             self.dt = self.options.dt
             if self.dt is None:
-                mesh_dt = self.eq_sw.get_time_step_advection(u_mag=self.options.u_advection)
-                dt = self.options.cfl_3d*float(np.floor(mesh_dt.dat.data.min()/20.0))
+                mesh_dt = self.compute_time_step_3d(u_mag=self.options.u_advection)
+                dt = self.options.cfl_3d*alpha*float(np.floor(mesh_dt.dat.data.min()))
```

Comment:
So alpha is a dt factor that depends on the time integration scheme and spatial discretization, here it's just implemented as a hacky number. In the ALE time integrator branch alpha is actually computed: all the time integrators define a CFL number and the element families have a scaling factor. The floor operator is a remnant of the past, it is removed in the ALE branch. I'd suggest that we leave this bit as it is because it will be revised in the (hopefully near) future.